### ISSUE TITLE
CMake option to turn off shared libraries

### ISSUE THREAD

**Issue by martinjrobins** [DrTimothyAldenDavis/SuiteSparse#424]
**Is your feature request related to a problem? Please describe.**
I'm interested in compiling the suitesparse KLU solver to webassembly using the wasi-sdk. This only supports building static libraries, so I would like to turn off building shared libraries

**Describe the solution you'd like**
A CMake option to turn off building shared libraries. E.g. `NSHARED=1` or `BUILD_SHARED_LIBS=OFF`

**Describe alternatives you've considered**
I've looked through the `CMakeLists.txt` under `KLU/` but could not see any way to turn off shared libs except by manually editing the file.

**Additional context**
If there is another way to turn off shared libs please let me know. Or alternatively I could put together a PR for the solution above.

**Comment by jingjingxyk** [comment]
fix static build :   https://github.com/jingjingxyk/SuiteSparse/pull/1/files


static build command (linux):


```bash

INSTALL_PREFIX=/usr/local/suitesparse
    
CMAKE_OPTIONS=" -DCMAKE_INSTALL_PREFIX=${INSTALL_PREFIX} "
CMAKE_OPTIONS="${CMAKE_OPTIONS} -DCMAKE_BUILD_TYPE=Release "
CMAKE_OPTIONS="${CMAKE_OPTIONS} -DBUILD_SHARED_LIBS=OFF "
CMAKE_OPTIONS="${CMAKE_OPTIONS} -DBUILD_STATIC_LIBS=ON "
CMAKE_OPTIONS="${CMAKE_OPTIONS} -DBLA_STATIC=ON "
CMAKE_OPTIONS="${CMAKE_OPTIONS} -DNSTATIC=OFF "
CMAKE_OPTIONS="${CMAKE_OPTIONS} -DDEMO=OFF "


# extra options

BLAS_INSTALL_PREFIX=/usr/local/blas
LAPACK_INSTALL_PREFIX=/usr/local/lapack
GMP_INSTALL_PREFIX=/usr/local/gmp
MPFR_INSTALL_PREFIX=/usr/local/mpfr

CMAKE_OPTIONS="${CMAKE_OPTIONS} -DCMAKE_PREFIX_PATH=\"${BLAS_INSTALL_PREFIX};${LAPACK_INSTALL_PREFIX};${GMP_INSTALL_PREFIX};${MPFR_INSTALL_PREFIX};\" ";


# build
make CMAKE_OPTIONS="'${CMAKE_OPTIONS}'" JOBS=$(nproc)

```

**Comment by mmuetzel** [comment]
That sounds like a reasonable suggestion to me.
If that is going to be implemented, it might be worth to consider switching to the "more standard" CMake variables `BUILD_SHARED_LIBS` and `BUILD_STATIC_LIBS` instead of `NSTATIC`. The latter might be a remnant of the prior Makefile build rules. 🤔
Imho, a goal of a potential implementation should be to still allow building shared *and* static libraries at the same time.

There are currently some changes to the CMake build system pending that would likely collide with these changes (e.g., adding a root CMakeLists.txt). It might be best to wait with this feature (support building only static libraries) after that (root CMakeLists.txt) is done.

**Comment by martinjrobins** [comment]
> There are currently some changes to the CMake build system pending that would likely collide with these changes (e.g., adding a root CMakeLists.txt). It might be best to wait with this feature (support building only static libraries) after that (root CMakeLists.txt) is done.

Sure, I'll wait until these changes have been merged and then put in the PR. Agree that `BUILD_STATIC_LIBS` is preferable to `NSTATIC`

**Comment by DrTimothyAldenDavis** [comment]
I think this is now fixed in SuiteSparse 7.4.0, in the default dev branch of the SuiteSparse repo.  I'll release a 7.4.0.beta1 soon, after finishing my own exhaustive coverage tests (not on github, but in SuiteSparse/<Package>/Tcov).

Can you give it a try?

**Comment by martinjrobins** [comment]
Thanks @DrTimothyAldenDavis, happy to report this works perfectly for me. I was able to complile `libklu.a` to wasm using emscripten (using openblas for the blas and lapack deps) and use it successfully within a sundials IDAS solver. Thank you for your help!
